File: runtime/dart/ui/gradient.ts
import { IList } from "../core/list";
import { pi } from "../math/pi";
import { IFloat64List } from "../typed_data/float64List";
import { IColor } from "./color";
import { IOffset } from "./offset";
import { IShader } from "./shader";
import { TileMode } from "./tileMode";
declare const dart: {
    ui: {
        gradientLinear: (
            from: IOffset,
            to: IOffset,
            colors: IList<IColor>,
            colorStops: IList<number> | undefined,
            tileMode: TileMode,
            matrix4?: IFloat64List | undefined
        ) => IGradient;
        gradientRadial: (
            center: IOffset,
            radius: number,
            colors: IList<IColor>,
            colorStops: IList<number> | undefined,
            tileMode: TileMode,
            matrix4: IFloat64List | undefined,
            focal: IOffset | undefined,
            focalRadius: number
        ) => IGradient;
        gradientSweep: (
            center: IOffset,
            colors: IList<IColor>,
            colorStops: IList<number> | undefined,
            tileMode: TileMode,
            startAngle: number,
            endAngle: number,
            matrix4?: IFloat64List | undefined
        ) => IGradient;
    };
};
export interface IGradient {
    toString: () => string;
    getHashCode: () => number;
}
export class Gradient implements IShader {
    public static linear(
        from: IOffset,
        to: IOffset,
        colors: IList<IColor>,
        colorStops: IList<number> | undefined,
        tileMode: TileMode = TileMode.clamp,
        matrix4?: IFloat64List | undefined
    ): IGradient {
        return dart.ui.gradientLinear(
            from,
            to,
            colors,
            colorStops,
            tileMode,
            matrix4
        );
    }
    public static radial(
        center: IOffset,
        radius: number,
        colors: IList<IColor>,
        colorStops: IList<number> | undefined,
        tileMode: TileMode = TileMode.clamp,
        matrix4: IFloat64List | undefined,
        focal: IOffset | undefined,
        focalRadius: number = 0.0
    ): IGradient {
        return dart.ui.gradientRadial(
            center,
            radius,
            colors,
            colorStops,
            tileMode,
            matrix4,
            focal,
            focalRadius
        );
    }
    public static sweep(
        center: IOffset,
        colors: IList<IColor>,
        colorStops: IList<number> | undefined,
        tileMode: TileMode = TileMode.clamp,
        startAngle: number = 0.0,
        endAngle: number = pi * 2,
        matrix4?: IFloat64List | undefined
    ): IGradient {
        return dart.ui.gradientSweep(
            center,
            colors,
            colorStops,
            tileMode,
            startAngle,
            endAngle,
            matrix4
        );
    }
    private readonly _dart_toString: () => string = undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    public toString(): string {
        return this._dart_toString();
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
}

File: lib/cfr/builtins/libs/dart/ui/gradient.dart
import 'dart:core';
import 'dart:typed_data';
import 'dart:ui';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedGradient extends VMManagedBox<Gradient> {
  VMManagedGradient(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final Gradient vmObject;
}

void loadGradient({required HydroState hydroState, required HydroTable table}) {
  table['gradientLinear'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Gradient>(
          object: Gradient.linear(
              maybeUnBoxAndBuildArgument<Offset, dynamic>(luaCallerArguments[1],
                  parentState: hydroState),
              maybeUnBoxAndBuildArgument<Offset, dynamic>(luaCallerArguments[2],
                  parentState: hydroState),
              maybeUnBoxAndBuildArgument<List<Color>, Color>(
                  luaCallerArguments[3],
                  parentState: hydroState),
              maybeUnBoxAndBuildArgument<List<double>?, double>(
                  luaCallerArguments[4],
                  parentState: hydroState),
              maybeUnBoxEnum(
                  values: TileMode.values, boxedEnum: luaCallerArguments[5]),
              maybeUnBoxAndBuildArgument<Float64List?, dynamic>(
                  luaCallerArguments[6],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['gradientRadial'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Gradient>(
          object: Gradient.radial(
              maybeUnBoxAndBuildArgument<Offset, dynamic>(luaCallerArguments[1],
                  parentState: hydroState),
              luaCallerArguments[2]?.toDouble(),
              maybeUnBoxAndBuildArgument<List<Color>, Color>(
                  luaCallerArguments[3],
                  parentState: hydroState),
              maybeUnBoxAndBuildArgument<List<double>?, double>(
                  luaCallerArguments[4],
                  parentState: hydroState),
              maybeUnBoxEnum(
                  values: TileMode.values, boxedEnum: luaCallerArguments[5]),
              maybeUnBoxAndBuildArgument<Float64List?, dynamic>(
                  luaCallerArguments[6],
                  parentState: hydroState),
              maybeUnBoxAndBuildArgument<Offset?, dynamic>(
                  luaCallerArguments[7],
                  parentState: hydroState),
              luaCallerArguments[8]?.toDouble()),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['gradientSweep'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Gradient>(
          object: Gradient.sweep(
              maybeUnBoxAndBuildArgument<Offset, dynamic>(luaCallerArguments[1],
                  parentState: hydroState),
              maybeUnBoxAndBuildArgument<List<Color>, Color>(
                  luaCallerArguments[2],
                  parentState: hydroState),
              maybeUnBoxAndBuildArgument<List<double>?, double>(
                  luaCallerArguments[3],
                  parentState: hydroState),
              maybeUnBoxEnum(
                  values: TileMode.values, boxedEnum: luaCallerArguments[4]),
              luaCallerArguments[5]?.toDouble(),
              luaCallerArguments[6]?.toDouble(),
              maybeUnBoxAndBuildArgument<Float64List?, dynamic>(
                  luaCallerArguments[7],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<Gradient>(boxer: (
      {required Gradient vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedGradient(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}
